### PR TITLE
initial work for iommu grouping

### PR DESCRIPTION
**Pull request by ibrokethecloud**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

There is a known issue while performing pci device passthrough when device in question has other devices in the same iommu group

**Solution:**

PR introduces an additional handler in the mutating webhook to patch VM's host devices.
If the VM contains a pcidevice which is part of a iommu group, then the webhook checks if other pcidevices which are part of the same iommu group have been enabled and added to the hostDevices section. 

If some devices are missing, then the webhook will create pcidevice claims for missing devices, if not present, and add them to the hostDevices list.

**Related Issue:**
https://github.com/harvester/harvester/issues/3451
**Test plan:**

### CODE REVIEW

File: pkg/util/fakeclients/pcidevicesclaim.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	pcidevicev1beta1 "github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
+	"github.com/harvester/pcidevices/pkg/generated/clientset/versioned/typed/devices.harvesterhci.io/v1beta1"
+	pcidevicesv1beta1ctl "github.com/harvester/pcidevices/pkg/generated/controllers/devices.harvesterhci.io/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+type PCIDevicesClaimClient func() v1beta1.PCIDeviceClaimInterface
```
[the remark this answers is not in the document]
Comment (ibrokethecloud) [reply pkg/util/fakeclients/pcidevicesclaim.go:15]: 👍

File: pkg/util/fakeclients/pcidevicesclaim.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"context"
+
+	pcidevicev1beta1 "github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
+	"github.com/harvester/pcidevices/pkg/generated/clientset/versioned/typed/devices.harvesterhci.io/v1beta1"
+	pcidevicesv1beta1ctl "github.com/harvester/pcidevices/pkg/generated/controllers/devices.harvesterhci.io/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
```
Comment (guangbochen) [inline pkg/util/fakeclients/pcidevicesclaim.go:12]: fix the import format

File: pkg/util/fakeclients/pcidevices.go [hunk 3]
```diff
@@ -12,6 +13,10 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 )
 
+const (
+	IommuGroupByNode = "harvesterhci.io/iommu-by-node"
```
[the remark this answers is not in the document]
Comment (ibrokethecloud) [reply pkg/util/fakeclients/pcidevices.go:17]: 👍

File: pkg/webhook/indexer.go [hunk 4]
```diff
@@ -4,20 +4,23 @@ import (
...
-	"github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
 	kubevirtv1 "kubevirt.io/api/core/v1"
+
+	"github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
 )
 
 const (
 	VMByName                = "harvesterhci.io/vm-by-name"
 	PCIDeviceByResourceName = "harvesterhcio.io/pcidevice-by-resource-name"
+	IommuGroupByNode        = "harvesterhci.io/iommu-by-node"
```
[the remark this answers is not in the document]
Comment (ibrokethecloud) [reply pkg/webhook/indexer.go:14]: 👍

File: pkg/util/fakeclients/pcidevicesclaim.go (hunk 2 above)
Comment (ibrokethecloud) [reply pkg/util/fakeclients/pcidevicesclaim.go:12]: 👍

File: pkg/webhook/vm_test.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+			Address:           "0000:04:10.0",
+			ClassId:           "0300",
+			Description:       "fake device 1",
+			NodeName:          "node2",
+			ResourceName:      "fake.com/device1",
+			VendorId:          "8086",
+			KernelDriverInUse: "ixgbevf",
+			IOMMUGroup:        "89",
+		},
+	}
```
Comment (tlehman) [inline pkg/webhook/vm_test.go:93]: Very thorough test data, nice

**Review by tlehman — APPROVED** [review]
This looks good except the API group, I'm going to merge it and change that instead of doing a complicated double fork.

**Pull request merged**